lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,538 @@
+    fn destruct(self) -> (usize, Vec<I>) {
+        (self.allocated_bytes, self.elements)
+    }
+
+    /// Whether the batch has space for a new item
+    ///
+    /// This function returns true of there is space both in terms of item count
+    /// and byte count for the given item, false otherwise.
+    fn has_space(&self, value: &I) -> bool {
+        let too_many_elements = self.elements.len() + 1 > self.element_limit;
+        let too_many_bytes = self.allocated_bytes + value.size_of() > self.allocation_limit;
+        !(too_many_elements || too_many_bytes)
+    }
+
+    /// Push an element into the batch
+    ///
+    /// This function pushes an element into the batch. Callers must be sure to
+    /// call `has_space` prior to calling this function and receive a positive
+    /// result.
+    ///
+    /// # Panics
+    ///
+    /// This function will panic if there is not sufficient space in the batch
+    /// for a new element to be inserted.
+    fn push(&mut self, value: I) {
+        assert!(self.has_space(&value));
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+    }
+}
+
+#[pin_project]
+pub struct Batcher<St, Prt, T>
+where
+    Prt: Partitioner,
+{
+    /// The timer to maintain periodic flushes from this `Batcher`
+    timer: T,
+    /// The maximum number of items that are allowed per-batch
+    batch_item_limit: usize,
+    /// The total number of bytes a single batch in this struct is allowed to
+    /// hold. If None effectively unlimited.
+    batch_allocation_limit: usize,
+    /// The store of live batches. Note that the key here is an option type,
+    /// on account of the interface of `Prt`.
+    batches: HashMap<Option<Prt::Key>, Batch<Prt::Item>, BuildHasherDefault<XxHash64>>,
+    /// The store of 'closed' batches. When this is not empty it will be
+    /// preferentially flushed prior to consuming any new items from the
+    /// underlying stream.
+    closed_batches: Vec<(Option<Prt::Key>, Vec<Prt::Item>)>,
+    /// The partitioner for this `Batcher`
+    partitioner: Prt,
+    #[pin]
+    /// The stream this `Batcher` wraps
+    stream: St,
+}
+
+impl<St, Prt, T> Batcher<St, Prt, T>
+where
+    St: Stream + Unpin,
+    Prt: Partitioner + Unpin,
+{
+    pub fn new(
+        stream: St,
+        partitioner: Prt,
+        timer: T,
+        batch_item_limit: NonZeroUsize,
+        batch_allocation_limit: Option<NonZeroUsize>,
+    ) -> Self {
+        Self {
+            batch_item_limit: batch_item_limit.get(),
+            batch_allocation_limit: batch_allocation_limit
+                .map_or(usize::max_value(), NonZeroUsize::get),
+            batches: HashMap::default(),
+            closed_batches: Vec::default(),
+            timer,
+            partitioner,
+            stream,
+        }
+    }
+}
+
+impl<St, Prt, T> Stream for Batcher<St, Prt, T>
+where
+    St: Stream + Unpin,
+    Prt: Partitioner + Unpin,
+    Prt::Item: ByteSizeOf + From<St::Item>,
```

Comment:
As @lukesteensen pointed out in chat, you can shift the item equality bounds from `Prt::Item` itself over to `St`:

    St: Stream<Item = Prt::Item> + Unpin,
    Prt: Partitioner + Unpin,
    Prt::Item: ByteSizeOf,

Reply:
So they're not quite equivalent, but I'm not sure it matters. The proposed change here requires the types to be the same -- which is what I originally wanted -- but I then worked myself around to believing that it'd be useful to get a conversion between types for free. Now I'm realizing if we _do_ want that a map would be better, so I'll switch this.

Reply:
Addressed in 0486830

Reply:
Makes sense, yeah.  I think it would end up being a little unclear if it could magically take anything `Into<Event>`, and as you said... it's a simple map if something actually needs to do that.